src/transformStateWithClones.js

```
@@ -1,13 +1,43 @@
 'use strict';
-
 /**
  * @param {Object} state
  * @param {Object[]} actions
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const count = [];
+
+  for (const n of actions) {
+    const lastState = count.length > 0 ? count[count.length - 1] : state;
+    let newState = { ...lastState };
+
+    switch (n.type) {
+      case 'addProperties':
+        Object.assign(newState, n.extraData);
+        break;
+
+      case 'removeProperties':
+        if (Array.isArray(n.keysToRemove)) {
+          for (const key of n.keysToRemove) {
+            delete newState[key];
+          }
+        } else {
+          return 'keysToRemove is not a Array';
```

This line returns a string when `n.keysToRemove` is not an array: `return 'keysToRemove is not a Array'`. That violates the task description requirement to “return an array with states recieved after each action.” Instead of returning a string, either throw an Error (preferred) or treat invalid `keysToRemove` as an empty array and continue. Returning a non-array breaks the function contract and will cause tests to fail.